core/storage/base_model/gae_models_test.py

```diff
@@ -236,6 +236,38 @@ def test_get_unversioned_instance_id(self):
         model1.put()
         self.assertEqual(model1.get_unversioned_instance_id(), 'model_id')
 
+    def test_export_data_trivial(self):
+        user_data = (base_models
+                     .BaseSnapshotMetadataModel
+                     .export_data('trivial_user'))
+        expected_data = {}
+        self.assertEqual(user_data, expected_data)
+
+    def test_export_data_nontrivial(self):
+        model1 = base_models.BaseSnapshotMetadataModel(
```

Comment:
In the codebase, BaseSnapshotMetadataModels are never created directly. Usually some subclass is called. I think you should mimic that (look at the codebase to see how the creation usually happens), and also have tests that make sure that stuff that's created in other subclasses or by other authors does not get picked up in this export method.

Reply:
Hey @seanlip, I was taking a look at this and I was wondering if you could point me in the right direction for how you want me to address this. I looked at the other tests in this file (for example, test_get_unversioned_instance_id) and these seem to directly create the model. Let me know how you want me to proceed.

Personally, my thinking is that we can directly create this model for this test, because if there are any problems when the model is created via a subclass, those problems will show up in the tests for those subclasses, and so we would catch them and handle them there.

Reply:
I think there are two ways:

(a) You could add tests to the respective subclass test files in such a way that they fully cover the stuff here, though that might be repetitive. 

(b) You could add a test for one representative subclass to this file.

I would still suggest that this method be tested by creating the model similar to how it is usually created in the non-test code, since this directly ties in to a user journey. Other tests that are already in this file might need to be revisited in the future for similar reasons.

Note also that if, as you say, the tests for the other subclasses already cover this code, then the above becomes trivial and you might not need to write anything new.

Does that help?

Reply:
Thanks for the response @seanlip. I think we're on the same page, but I want to make sure. 

We will inevitably have to implement tests for all of the subclasses by virtue of how we are currently testing the full takeout functionality in core.domain.takeout_service. Thus, I think that for now having this level of testing is fine (and seems to match the testing in the rest of the file), and then testing for the respective subclasses will happen at the domain level (alongside the rest of takeout functionality testing). 

Let me know if this makes sense or I'm misunderstanding something.

Reply:
No, that's not quite right. This PR and the discussion above cover the storage level so all storage-level tests should go in this PR. In the response above, I am not talking about domain-level tests.

Is there a reason why we can't just change the storage level model used here to be one of the subclass models?

Reply:
So you mean just pick any storage level subclass and swap it out with the current code?

Reply:
What do you mean by "swap it out with the current code"?

(I meant what I said in https://github.com/oppia/oppia/pull/8749#discussion_r388666296)

Reply:
When you say https://github.com/oppia/oppia/pull/8749#discussion_r388667845

Do you mean taking this line:

`model1 = base_models.BaseSnapshotMetadataModel(
            id='model_id-1', committer_id='committer_id', commit_type='create')`

and replacing BaseSnapshotMetadataModel with a subclass.

Reply:
Take a look at base_model/gae_models.py (line 693) to see how this class is typically created. Follow a similar invocation in your tests using a specific subclass.

Reply:
@seanlip I think I addressed this in my recent commit. Let me know if I misunderstood!